openstack/pkg/designate/client.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		return "", err
+	}
+
+	for _, rr := range allRecordSets {
+		if recordName == rr.Name {
+			return rr.ID, nil
+		}
+	}
+
+	return "", nil
+}
+
+func (p *Provider) setRecordID(recordID string) {
+	p.recordID = recordID
+}
+
+func (p *Provider) setDescription(desc string) error {
+	if desc == "" {
+		p.dnsDescription = "example description"
+	}
+	p.dnsDescription = desc
+
+	return nil
+}
+
+func (p *Provider) createRecord(record libdns.Record, zone string) error {
+	createOpts := recordsets.CreateOpts{
+		Name:    record.Name + zone,
+		Type:    record.Type,
+		TTL:     int(record.TTL / time.Second),
+		Records: []string{record.Value},
+	}
+
+	exist, err := p.getRecordID(record.Name, zone)
+	if err != nil {
+		return fmt.Errorf("cannot get recordID: %v", err)
+	}
+
+	if exist != "" {
+		return errors.New("DNS record already exist")
+	}
+
+	_, err = recordsets.Create(p.dnsClient, p.zoneID, createOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("cannot create DNS record: %v", err)
+	}
+
+	return nil
+}
+
+func (p *Provider) updateRecord(record libdns.Record, zone string) error {
+	updateOpts := recordsets.UpdateOpts{
+		TTL:     intToPointer(int(record.TTL / time.Second)),
+		Records: []string{record.Value},
+	}
+
+	// Update updates a recordset in a given zone
+	_, err := recordsets.Update(p.dnsClient, p.zoneID, p.recordID, updateOpts).Extract()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Provider) deleteRecord(record libdns.Record, zone string) error {
+	err := recordsets.Delete(p.dnsClient, p.zoneID, p.recordID).ExtractErr()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Provider) exportEnvVariables() error {
+	err := os.Setenv("OS_REGION_NAME", p.AuthOpenStack.RegionName)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_TENANT_ID", p.AuthOpenStack.TenantID)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_IDENTITY_API_VERSION", p.AuthOpenStack.IdentityApiVersion)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_PASSWORD", p.AuthOpenStack.Password)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_AUTH_URL", p.AuthOpenStack.AuthURL)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_USERNAME", p.AuthOpenStack.Username)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_TENANT_NAME", p.AuthOpenStack.TenantName)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_ENDPOINT_TYPE", p.AuthOpenStack.EndpointType)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	return nil
+}
+
+func (p *Provider) auth(zoneName string) error {
```

Comment:
This isn't thread-safe, client could authenticate multiple times.

Reply:
I added small check if client is already authenticated.